File: app/src/main/java/com/proxerme/app/adapter/NewsAdapter.java
package com.proxerme.app.adapter;

import android.os.Bundle;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.mikepenz.community_material_typeface_library.CommunityMaterial;
import com.mikepenz.iconics.IconicsDrawable;
import com.proxerme.app.R;
import com.proxerme.app.util.TimeUtils;
import com.proxerme.library.connection.UrlHolder;
import com.proxerme.library.entity.News;
import com.proxerme.library.util.ProxerInfo;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * An adapter for {@link News}, for usage in a {@link RecyclerView}.
 *
 * @author Ruben Gees
 */
public class NewsAdapter extends PagingAdapter<News, NewsAdapter.ViewHolder> {

    private static final String STATE_NEWS_EXTENSION_IDS = "news_extension_ids";
    private static final int ICON_SIZE = 32;
    private static final int ICON_PADDING = 8;
    private static final float ROTATION_HALF = 180f;
    private static final int DESCRIPTION_MAX_LINES = 3;

    private HashMap<String, Boolean> extensionMap;

    private OnNewsInteractionListener onNewsInteractionListener;

    public NewsAdapter() {
        super();
        extensionMap = new HashMap<>(ProxerInfo.NEWS_ON_PAGE * 2);
    }

    public NewsAdapter(@NonNull Collection<News> news) {
        super(news);
        extensionMap = new HashMap<>(news.size() * 2);
    }

    public NewsAdapter(@NonNull Bundle savedInstanceState) {
        super(savedInstanceState);
        List<String> ids = savedInstanceState.getStringArrayList(STATE_NEWS_EXTENSION_IDS);
        extensionMap = new HashMap<>();

        if (ids != null) {
            for (String id : ids) {
                extensionMap.put(id, true);
            }
        }
    }

    @Override
    public NewsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_news, parent, false));
    }

    @Override
    public void onBindViewHolder(NewsAdapter.ViewHolder holder, int position) {
        News item = getItemAt(position);

        holder.title.setText(item.getSubject().trim());
        holder.description.setText(item.getDescription().trim());
        holder.category.setText(item.getCategoryTitle());
        holder.time.setText(TimeUtils.convertToRelativeReadableTime(holder.time.getContext(),
                item.getTime()));

        holder.expand.setImageDrawable(new IconicsDrawable(holder.expand.getContext())
                .colorRes(R.color.icons_grey).sizeDp(ICON_SIZE).paddingDp(ICON_PADDING)
                .icon(CommunityMaterial.Icon.cmd_chevron_down));

        if (extensionMap.containsKey(item.getId())) {
            holder.description.setMaxLines(Integer.MAX_VALUE);
            ViewCompat.setRotationX(holder.expand, ROTATION_HALF);
        } else {
            holder.description.setMaxLines(DESCRIPTION_MAX_LINES);
            ViewCompat.setRotationX(holder.expand, 0f);
        }

        Glide.with(holder.image.getContext()).load(UrlHolder.getNewsImageUrl(item.getId(),
                item.getImageId())).diskCacheStrategy(DiskCacheStrategy.SOURCE).into(holder.image);
    }

    @Override
    public void saveInstanceState(@NonNull Bundle outState) {
        super.saveInstanceState(outState);
        outState.putStringArrayList(STATE_NEWS_EXTENSION_IDS,
                new ArrayList<>(extensionMap.keySet()));
    }

    @Override
    @IntRange(from = 1)
    protected int getItemsOnPage() {
        return ProxerInfo.NEWS_ON_PAGE;
    }

    public void setOnNewsInteractionListener(@Nullable OnNewsInteractionListener
                                                     onNewsInteractionListener) {
        this.onNewsInteractionListener = onNewsInteractionListener;
    }

    public static abstract class OnNewsInteractionListener {
        public void onNewsClick(@NonNull View v, @NonNull News news) {

        }

        public void onNewsImageClick(@NonNull View v, @NonNull News news) {

        }

        public void onNewsExpanded(@NonNull View v, @NonNull News news) {

        }
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        @Bind(R.id.item_news_image)
        ImageView image;
        @Bind(R.id.item_news_title)
        TextView title;
        @Bind(R.id.item_news_description)
        TextView description;
        @Bind(R.id.item_news_category)
        TextView category;
        @Bind(R.id.item_news_time)
        TextView time;
        @Bind(R.id.item_news_expand_description)
        ImageButton expand;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @OnClick(R.id.item_news_content_container)
        void onContentClick(View view) {
            if (onNewsInteractionListener != null) {
                onNewsInteractionListener.onNewsClick(view, getItemAt(getAdapterPosition()));
            }
        }

        @OnClick(R.id.item_news_image)
        void onImageClick(View view) {
            if (onNewsInteractionListener != null) {
                onNewsInteractionListener.onNewsImageClick(view, getItemAt(getAdapterPosition()));
            }
        }

        @OnClick(R.id.item_news_expand_description)
        void onExpandClick(View view) {
            News news = getItemAt(getAdapterPosition());
            String id = news.getId();
            boolean isExpanded = extensionMap.containsKey(id);

            if (isExpanded) {
                extensionMap.remove(id);

                description.setMaxLines(DESCRIPTION_MAX_LINES);
                ViewCompat.animate(view).rotationX(0f);
            } else {
                extensionMap.put(id, true);

                description.setMaxLines(Integer.MAX_VALUE);
                ViewCompat.animate(view).rotationX(ROTATION_HALF);

                if (onNewsInteractionListener != null) {
                    onNewsInteractionListener.onNewsExpanded(view, news);
                }
            }
        }
    }
}

File: app/src/main/kotlin/com/proxerme/app/adapter/NewsAdapter.kt
package com.proxerme.app.adapter

import android.os.Bundle
import android.support.v4.view.ViewCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import butterknife.bindView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.mikepenz.community_material_typeface_library.CommunityMaterial
import com.mikepenz.iconics.IconicsDrawable
import com.proxerme.app.R
import com.proxerme.app.util.TimeUtil
import com.proxerme.library.connection.notifications.entitiy.News
import com.proxerme.library.info.ProxerUrlHolder
import java.util.*

/**
 * TODO: Describe Class
 *
 * @author Ruben Gees
 */
class NewsAdapter(savedInstanceState: Bundle? = null) :
        PagingAdapter<News>() {

    private companion object {
        private const val ITEMS_STATE = "adapter_news_state_items"
        private const val EXPANDED_IDS_STATE = "adapter_news_state_extension_ids"
        private const val ICON_SIZE = 32
        private const val ICON_PADDING = 8
        private const val ROTATION_HALF = 180f
        private const val DESCRIPTION_MAX_LINES = 3
    }

    var callback: OnNewsInteractionListener? = null

    private val expanded = HashMap<String, Boolean>()

    init {
        savedInstanceState?.let {
            list.addAll(it.getParcelableArrayList(ITEMS_STATE))
            it.getStringArrayList(EXPANDED_IDS_STATE)
                    .associateByTo(expanded, { it }, { true })
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder? {
        return ViewHolder(LayoutInflater.from(parent.context)
                .inflate(R.layout.item_news, parent, false))
    }

    override fun saveInstanceState(outState: Bundle) {
        outState.putParcelableArrayList(ITEMS_STATE, list)
        outState.putStringArrayList(EXPANDED_IDS_STATE, ArrayList(expanded.keys))
    }

    inner class ViewHolder(itemView: View) : PagingAdapter.PagingViewHolder<News>(itemView) {

        private val expand: ImageButton by bindView(R.id.expand)
        private val description: TextView by bindView(R.id.description)
        private val image: ImageView by bindView(R.id.image)
        private val title: TextView by bindView(R.id.title)
        private val category: TextView by bindView(R.id.category)
        private val time: TextView by bindView(R.id.time)

        init {
            expand.setImageDrawable(IconicsDrawable(expand.context)
                    .colorRes(R.color.icon)
                    .sizeDp(ICON_SIZE)
                    .paddingDp(ICON_PADDING)
                    .icon(CommunityMaterial.Icon.cmd_chevron_down))

            expand.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    val id = list[adapterPosition].id

                    if (expanded.containsKey(id)) {
                        expanded.remove(id)

                        description.maxLines = DESCRIPTION_MAX_LINES
                        ViewCompat.animate(expand).rotation(0f)
                    } else {
                        expanded.put(id, true)

                        description.maxLines = Integer.MAX_VALUE
                        ViewCompat.animate(expand).rotation(ROTATION_HALF)
                    }

                    callback?.onNewsExpanded(it, list[adapterPosition])
                }
            }

            itemView.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    callback?.onNewsClick(it, list[adapterPosition])
                }
            }

            image.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    callback?.onNewsImageClick(it, list[adapterPosition])
                }
            }
        }

        override fun bind(item: News) {
            Glide.with(image.context)
                    .load(ProxerUrlHolder.getNewsImageUrl(item.id, item.imageId).toString())
                    .diskCacheStrategy(DiskCacheStrategy.SOURCE)
                    .into(image)

            title.text = item.subject
            description.text = item.description
            category.text = item.categoryTitle
            time.text = TimeUtil.convertToRelativeReadableTime(time.context,
                    item.time)

            if (expanded.containsKey(item.id)) {
                description.maxLines = Integer.MAX_VALUE
                ViewCompat.setRotation(expand, ROTATION_HALF)
            } else {
                description.maxLines = DESCRIPTION_MAX_LINES
                ViewCompat.setRotation(expand, 0f)
            }
        }
    }

    abstract class OnNewsInteractionListener {
        open fun onNewsClick(v: View, news: News) {

        }

        open fun onNewsImageClick(v: View, news: News) {

        }

        open fun onNewsExpanded(v: View, news: News) {

        }
    }
}